From 2cae93832f46b245847bdc252456ddf7742ef45e Mon Sep 17 00:00:00 2001
From: Sebastian Pipping <sebastian@pipping.org>
Date: Mon, 21 Feb 2022 15:48:32 +0100
Subject: [PATCH] bpo-46811: Make test suite support Expat >=2.4.5 (GH-31453)

Curly brackets were never allowed in namespace URIs
according to RFC 3986, and so-called namespace-validating
XML parsers have the right to reject them a invalid URIs.

libexpat >=2.4.5 has become strcter in that regard due to
related security issues; with ET.XML instantiating a
namespace-aware parser under the hood, this test has no
future in CPython.

References:
- https://datatracker.ietf.org/doc/html/rfc3968
- https://www.w3.org/TR/xml-names/

Also, test_minidom.py: Support Expat >=2.4.5

Backport:
 * Drop Misc/NEWS.d
 * test_minidom.py: Update import context
 * test_minidom.py: Drop byte string prefixes
 * test_minidom.py: Remove expat version check as the expat changes have been
   backported in Debian.
 * test_minidom.py: Drop testExceptionOnSpacesInXMLNSValue hunk

---
 Lib/test/test_minidom.py                        | 17 +++++++++++++++--
 Lib/test/test_xml_etree.py                      |  6 ------
 .../2022-02-20-21-03-31.bpo-46811.8BxgdQ.rst    |  1 +
 3 files changed, 16 insertions(+), 8 deletions(-)
 create mode 100644 Misc/NEWS.d/next/Library/2022-02-20-21-03-31.bpo-46811.8BxgdQ.rst

diff --git a/Lib/test/test_minidom.py b/Lib/test/test_minidom.py
index 1663b1f114..97620258d8 100644
--- a/Lib/test/test_minidom.py
+++ b/Lib/test/test_minidom.py
@@ -6,12 +6,14 @@
 from test import support
 import unittest
 
+import pyexpat
 import xml.dom
 import xml.dom.minidom
 import xml.parsers.expat
 
 from xml.dom.minidom import parse, Node, Document, parseString
 from xml.dom.minidom import getDOMImplementation
+from xml.parsers.expat import ExpatError
 
 
 tstfile = support.findfile("test.xml", subdir="xmltestdata")
@@ -1147,8 +1149,10 @@ def testEncodings(self):
 
         # Verify that character decoding errors raise exceptions instead
         # of crashing
-        self.assertRaises(UnicodeDecodeError, parseString,
-                '<fran\xe7ais>Comment \xe7a va ? Tr\xe8s bien ?</fran\xe7ais>')
+        self.assertRaises(ExpatError, parseString,
+                '<fran\xe7ais></fran\xe7ais>')
+        self.assertRaises(ExpatError, parseString,
+                '<franais>Comment \xe7a va ? Tr\xe8s bien ?</franais>')

         doc.unlink()

diff --git a/Lib/test/test_xml_etree.py b/Lib/test/test_xml_etree.py
index a25f536134..c5292b5e9e 100644
--- a/Lib/test/test_xml_etree.py
+++ b/Lib/test/test_xml_etree.py
@@ -2192,12 +2192,6 @@ def test_issue6233(self):
                 b"<?xml version='1.0' encoding='ascii'?>\n"
                 b'<body>t&#227;g</body>')
 
-    def test_issue3151(self):
-        e = ET.XML('<prefix:localname xmlns:prefix="${stuff}"/>')
-        self.assertEqual(e.tag, '{${stuff}}localname')
-        t = ET.ElementTree(e)
-        self.assertEqual(ET.tostring(e), b'<ns0:localname xmlns:ns0="${stuff}" />')
-
     def test_issue6565(self):
         elem = ET.XML("<body><tag/></body>")
         self.assertEqual(summarize_list(elem), ['tag'])
